components/usage/pkg/controller/reconciler.go

```diff
@@ -108,129 +102,38 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
-
-	return status, nil
-}
-
-func generateUsageReport(teams []teamWithWorkspaces, maxStopTime time.Time) ([]TeamUsage, error) {
-	var report []TeamUsage
-	for _, team := range teams {
-		var teamTotalRuntime int64
-		for _, workspace := range team.Workspaces {
-			for _, instance := range workspace.Instances {
-				teamTotalRuntime += instance.WorkspaceRuntimeSeconds(maxStopTime)
-			}
-		}
+	instancesByAttributionID := groupInstancesByAttributionID(instances)
 
-		report = append(report, TeamUsage{
-			TeamID:           team.TeamID.String(),
-			WorkspaceSeconds: teamTotalRuntime,
-		})
-	}
-	return report, nil
-}
+	u.billingController.Reconcile(ctx, now, instancesByAttributionID)
 
-type teamWithWorkspaces struct {
-	TeamID     uuid.UUID
-	Workspaces []workspaceWithInstances
+	return status, instancesByAttributionID, nil
 }
 
-func (u *UsageReconciler) loadTeamsForWorkspaces(ctx context.Context, workspaces []workspaceWithInstances) ([]teamWithWorkspaces, error) {
-	// find owner IDs of these workspaces
-	var ownerIDs []uuid.UUID
-	for _, workspace := range workspaces {
-		ownerIDs = append(ownerIDs, workspace.Workspace.OwnerID)
-	}
+type UsageReport map[db.AttributionID][]db.WorkspaceInstance
 
-	// Retrieve memberships. This gives a link between an Owner and a Team they belong to.
-	memberships, err := db.ListTeamMembershipsForUserIDs(ctx, u.conn, ownerIDs)
-	if err != nil {
-		return nil, fmt.Errorf("failed to list team memberships: %w", err)
-	}
+func (u UsageReport) RuntimeSummaryForTeams(maxStopTime time.Time) map[string]int64 {
+	attributedUsage := map[string]int64{}
 
-	membershipsByUserID := map[uuid.UUID]db.TeamMembership{}
-	for _, membership := range memberships {
-		// User can belong to multiple teams. For now, we're choosing the membership at random.
-		membershipsByUserID[membership.UserID] = membership
-	}
+	for attribution, instances := range u {
+		entity, id := attribution.Values()
+		if entity != db.AttributionEntity_Team {
+			continue
+		}
 
-	// Convert workspaces into a lookup so that we can index into them by Owner ID, needed for joining Teams with Workspaces
-	workspacesByOwnerID := map[uuid.UUID][]workspaceWithInstances{}
-	for _, workspace := range workspaces {
-		workspacesByOwnerID[workspace.Workspace.OwnerID] = append(workspacesByOwnerID[workspace.Workspace.OwnerID], workspace)
-	}
+		var runtime uint64
+		for _, instance := range instances {
+			runtime += instance.WorkspaceRuntimeSeconds(maxStopTime)
+		}
 
-	// Finally, join the datasets
-	// Because we iterate over memberships, and not workspaces, we're in effect ignoring Workspaces which are not in a team.
-	// This is intended as we focus on Team usage for now.
-	var teamsWithWorkspaces []teamWithWorkspaces
-	for userID, membership := range membershipsByUserID {
-		teamsWithWorkspaces = append(teamsWithWorkspaces, teamWithWorkspaces{
-			TeamID:     membership.TeamID,
-			Workspaces: workspacesByOwnerID[userID],
-		})
+		attributedUsage[id] = int64(runtime)
```

Comment:
This feels dangerous. Why not make attributedUsage accumulate in `uint64`?

Reply:
Happy to update to unit64 everyone (but Stripe only accepts int64, not unit64). But the reason for using it here is to limit changes to the existing Usage Report reconcile logic with stripe in this PR. That happens here https://github.com/gitpod-io/gitpod/pull/10938/files#diff-b2499b7086d5733f081dcce586e0ff0e77206dd5d8fd6ede638e3fc8f284c798L25-L32 and it currently has int64 as the interface.
I'll update these in a follow-up PR (minus the Stripe API call which will have to be int64) if that's acceptable.

Reply:
Ok, fine with this for now.